presto-native-sidecar-plugin/src/main/java/com/facebook/presto/sidecar/nativechecker/NativePlanChecker.java

```diff
@@ -0,0 +1,231 @@
+            }
+        }
+        catch (final IOException e) {
+            throw new PrestoException(NATIVEPLANCHECKER_CONNECTION_ERROR, "I/O error getting native plan checker response", e);
+        }
+        finally {
+            LOG.debug("Native plan validation complete [fragment: %s, root: %s]", planFragment.getId(), planFragment.getRoot().getId());
+        }
+    }
+
+    private Request buildRequest(String requestBodyJson)
+    {
+        // Use native sidecar plan conversion endpoint to validate
+        String planConversionUrl = nodeManager.getSidecarNode().getHttpUri().toString() + PLAN_CONVERSION_ENDPOINT;
+
+        Request.Builder builder = new Request.Builder()
+                .url(planConversionUrl)
+                .addHeader("CONTENT_TYPE", "APPLICATION_JSON")
+                .post(RequestBody.create(JSON_CONTENT_TYPE, requestBodyJson));
+
+        return builder.build();
+    }
+
+    private PlanConversionFailure processResponseFailure(Response response) throws IOException
+    {
+        if (response.body() == null) {
+            throw new PrestoException(NATIVEPLANCHECKER_UNKNOWN_CONVERSION_FAILURE, "Error response without failure from native plan checker with code: " + response.code());
+        }
+
+        PlanConversionResponse planConversionResponse = PLAN_CONVERSION_RESPONSE_JSON_CODEC.fromJson(response.body().bytes());
+        if (planConversionResponse.getFailures().isEmpty()) {
+            throw new PrestoException(NATIVEPLANCHECKER_UNKNOWN_CONVERSION_FAILURE, "Error response without failure from native plan checker with code: " + response.code());
+        }
+
+        return planConversionResponse.getFailures().get(0);
+    }
+
+    private static class CheckInternalVisitor
+            extends PlanVisitor<Boolean, Void>
+    {
+        @Override
+        public Boolean visitTableScan(TableScanNode tableScan, Void context)
+        {
+            TableHandle handle = tableScan.getTable();
+            return ConnectorId.isInternalSystemConnector(handle.getConnectorId());
+        }
+
+        @Override
+        public Boolean visitPlan(PlanNode node, Void context)
+        {
+            for (PlanNode child : node.getSources()) {
+                if (child.accept(this, context)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+    }
+
+    public static class PlanConversionResponse
+    {
+        private final List<PlanConversionFailure> failures;
+
+        @JsonCreator
+        public PlanConversionResponse(
+                @JsonProperty("failures") List<PlanConversionFailure> failures)
+        {
+            this.failures = ImmutableList.copyOf(requireNonNull(failures, "failures is null"));
+        }
+
+        @JsonProperty
+        public List<PlanConversionFailure> getFailures()
+        {
+            return failures;
+        }
+    }
+
+    public static class PlanConversionFailure
```

Comment:
I think I'd love some examples of what this will look like in various failure scenarios to understand this better.

Reply:
I'm happy to try and trigger whatever failures you like, but if you're wondering how failure info is generated the steps are:

1. plain conversion fails with an exception
2. the exception is cause and translated to `ExceptionFailureInfo` at https://github.com/prestodb/presto/blob/master/presto-native-execution/presto_cpp/main/common/Exception.cpp#L18
3. the failure info is added to the response

So the translation code has a few hardcoded values, and some depend on what is available in the exception